File: 0109_Convert_Sorted_List_to_Binary_Search_Tree/solution.py
'''
109. Convert Sorted List to Binary Search Tree

Given a singly linked list where elements are sorted in ascending order, convert it to a height balanced BST.
'''
from data_structure.tree_node import TreeNode
from utilities.data_generator import *


class Solution(object):
    def sortedListToBST(self, head):
        if not head:
           return None
        if not head.next:
            return TreeNode(head.val)

        fast = head.next
        slow = head
        last_of_left = None
        while fast and fast.next:
            fast = fast.next.next
            last_of_left = slow
            slow = slow.next

        root = TreeNode(slow.val)

        if last_of_left:
            last_of_left.next = None
            root.left = self.sortedListToBST(head)

        if slow.next:
            root.right = self.sortedListToBST(slow.next)
        return root


    def sortedListToBST_2(self, head):
        """
        :type head: ListNode
        :rtype: TreeNode
        """
        self.cur = head
        node, length = head, 0
        while node:
            node = node.next
            length += 1
        return self.build_tree(length)

    def build_tree(self, n):
        if n <= 0:
            return None
        l = self.build_tree(n / 2)
        root = TreeNode(self.cur.val)
        self.cur = self.cur.next
        r = self.build_tree(n - n / 2 - 1)
        root.left = l
        root.right = r
        return root


s = Solution()
head = generate_list([1, 2, 3])
tree = s.sortedListToBST_2(head)

File: 0109_Convert_Sorted_List_to_Binary_Search_Tree/solution.cpp
/*
109. Convert Sorted List to Binary Search Tree

Level: Medium

https://leetcode.com/problems/convert-sorted-list-to-binary-search-tree
*/
#include <iostream>
#include <vector>
#include <string>
#include <stack>
#include <queue>
#include <unordered_map>
#include <unordered_set>
#include <numeric>
#include <algorithm>

#include "common/ListNode.h"
#include "common/TreeNode.h"

using namespace std;

/*
  Solution: 
*/

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution
{
public:
  TreeNode *sortedListToBST(ListNode *head)
  {
    if (head == nullptr)
    {
      return nullptr;
    }

    ListNode *end = head;
    while (end->next != nullptr)
    {
      end = end->next;
    }

    return sortedListToBSTRecursion(head, end);
  }

private:
  TreeNode *sortedListToBSTRecursion(ListNode *head, ListNode *end)
  {
    if (head == nullptr || end == nullptr || head == end->next)
    {
      return nullptr;
    }

    ListNode *slow = head;
    ListNode *preslow = nullptr;
    ListNode *fast = head;
    while (fast != end && fast->next != end)
    {
      preslow = slow;
      slow = slow->next;
      fast = fast->next->next;
    }

    TreeNode *root = new TreeNode(slow->val);
    root->left = sortedListToBSTRecursion(head, preslow);
    root->right = sortedListToBSTRecursion(slow->next, end);

    return root;
  }
};